File: runtime/elixir/lib/runtime_utils.ex
defmodule RuntimeUtils do
  @moduledoc false

  @spec split_base_url(String.t()) :: {String.t(), String.t()}
  def split_base_url(""), do: {"", ""}

  def split_base_url(base_url) do
    case URI.parse(base_url) do
      %URI{scheme: scheme, host: host} = uri when is_binary(host) ->
        port_suffix =
          case {uri.scheme, uri.port} do
            {"https", 443} -> ""
            {"http", 80} -> ""
            {_, nil} -> ""
            {_, port} -> ":#{port}"
          end

        {scheme <> "://", host <> port_suffix}

      _ ->
        {"", base_url}
    end
  end

  @spec endpoint_host_from_config(map()) :: String.t() | nil
  def endpoint_host_from_config(config) do
    case Map.get(config, :base_url) do
      nil ->
        case {Map.get(config, :endpoint_prefix), Map.get(config, :region, "us-east-1")} do
          {nil, _} -> nil
          {prefix, region} -> "#{prefix}.#{region}.amazonaws.com"
        end

      base_url ->
        {_scheme, authority} = split_base_url(base_url)
        authority
    end
  end
end

File: runtime/erlang/src/runtime_utils.erl
%% Shared smithy-beam Erlang runtime helpers.
-module(runtime_utils).
-export([
    split_base_url/1,
    endpoint_host_from_config/1
]).

%% @doc Split a base URL into scheme prefix and authority.
-spec split_base_url(binary()) -> {binary(), binary()}.
split_base_url(<<>>) ->
    {<<>>, <<>>};
split_base_url(BaseUrl) ->
    case uri_string:parse(binary_to_list(BaseUrl)) of
        #{scheme := Scheme, host := Host} = Parts ->
            PortSuffix =
                case maps:get(port, Parts, undefined) of
                    undefined -> <<>>;
                    Port -> <<":", (integer_to_binary(Port))/binary>>
                end,
            {<<(list_to_binary(Scheme))/binary, "://">>, <<
                (list_to_binary(Host))/binary, PortSuffix/binary
            >>};
        _ ->
            {<<>>, BaseUrl}
    end.

%% @doc Resolve the HTTP host from client config or AWS endpoint metadata.
-spec endpoint_host_from_config(#{binary() => term()}) -> binary() | undefined.
endpoint_host_from_config(Config) ->
    case maps:get(base_url, Config, undefined) of
        undefined ->
            case
                {
                    maps:get(endpoint_prefix, Config, undefined),
                    maps:get(region, Config, <<"us-east-1">>)
                }
            of
                {undefined, _} -> undefined;
                {Prefix, Region} -> <<Prefix/binary, ".", Region/binary, ".amazonaws.com">>
            end;
        BaseUrl ->
            {_Scheme, Authority} = split_base_url(BaseUrl),
            Authority
    end.
